extensions/amp-viewer-integration/0.1/test/test-amp-viewer-integration.js

```diff
@@ -14,9 +14,193 @@
  * limitations under the License.
  */
 
+import {Messaging} from '../messaging.js';
+import {Viewer} from './viewer-messaging-example.js';
+import {adopt} from '../../../../src/runtime';
+import * as sinon from 'sinon';
+
+
+adopt(window);
 
 describes.sandboxed('AmpViewerIntegration', {}, () => {
-  it('should work', () => {
-    // TODO
+  const ampDocSrc = '/test/fixtures/served/ampdoc-with-messaging.html';
+
+  let viewerEl;
+  let viewer;
+
+  beforeEach(() => {
+    const loc = window.location;
+    const ampDocUrl =
+      `${loc.protocol}//iframe.${loc.hostname}:${loc.port}${ampDocSrc}`;
+
+    viewerEl = document.createElement('div');
+    document.body.appendChild(viewerEl);
+    viewer = new Viewer(viewerEl, '1', ampDocUrl, true);
+    return viewer.waitForHandshakeRequest();
+  });
+
+  afterEach(() => {
+    document.body.removeChild(viewerEl);
+  });
+
+  describe('Handshake', function() {
+    this.timeout(5000);
+    it('should confirm the handshake', () => {
+      console.log('sending handshake response');
+      viewer.confirmHandshake();
+      return viewer.waitForDocumentLoaded();
+    });
+  });
+
+  describe('Unit Tests', () => {
+    const viewerOrigin = 'http://localhost:9876';
+    const ampDoc = 'http://localhost:8000/examples/everything.amp.max.html';
+    const requestProcessor = function() {};
+    let messaging;
+
+    beforeEach(() => {
+      const source = {
+        postMessage: function() {},
+        addEventListener: function() {},
+      };
+      messaging = new Messaging(
+        source, window, viewerOrigin, requestProcessor, ampDoc);
+    });
+
+    afterEach(() => {
+      messaging = null;
+    });
+
+    it('should initialize correctly', () => {
+      expect(messaging.requestSentinel_).to.equal('__AMPHTML__REQUEST');
+      expect(messaging.responseSentinel_).to.equal('__AMPHTML__RESPONSE');
+      expect(messaging.requestProcessor_).to.equal(requestProcessor);
+      expect(messaging.target_).to.equal(window);
+      expect(messaging.targetOrigin_).to.equal(viewerOrigin);
+    });
+
+    it('handleMessage_ should call handleRequest_', () => {
```

Comment:
Any chance these tests could be rewritten to be much higher level?

I'd recommend to get rid of all of these tests that do stubbing and expect methods to be called.

Instead make a fake window and do assertion about the postMessages being called against the parent and for incoming events to do the right thing.